From fec35db7e02a5a9345c3be22b07325b4cf2e96c4 Mon Sep 17 00:00:00 2001
From: Chen Baozi <chenbaozi@phytium.com.cn>
Date: Mon, 24 Jun 2024 21:17:22 +0800
Subject: [PATCH 141/150] net: macb: Unregister fixed clock defined in platform
 data

We manually register fixed 'pclk' and 'hclk' for ACPI which require to
be freed explicitly when the driver is removed. Otherwise, the driver
module cannot be reinitialized and return -EEXIST error code.

Signed-off-by: Chen Baozi <chenbaozi@phytium.com.cn>
Change-Id: Iea268aa1afe694a12614ffcc41c034b31e302829
Signed-off-by: Andrew Powers-Holmes <aholmes@omnom.net>
---
 drivers/net/ethernet/cadence/macb_main.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/cadence/macb_main.c b/drivers/net/ethernet/cadence/macb_main.c
index a06b996c0116..85f55330193b 100644
--- a/drivers/net/ethernet/cadence/macb_main.c
+++ b/drivers/net/ethernet/cadence/macb_main.c
@@ -5073,10 +5073,10 @@ static int phytium_clk_init(struct platform_device *pdev, struct clk **pclk,
 	return 0;
 
 err_plat_dev_register:
-	clk_unregister(plat_data.hclk);
+	clk_unregister_fixed_rate(plat_data.hclk);
 
 err_hclk_register:
-	clk_unregister(plat_data.pclk);
+	clk_unregister_fixed_rate(plat_data.pclk);
 
 err_pclk_register:
 	return err;
@@ -5598,6 +5598,7 @@ static int macb_remove(struct platform_device *pdev)
 {
 	struct net_device *dev;
 	struct macb *bp;
+	struct macb_platform_data *pdata;
 
 	dev = platform_get_drvdata(pdev);
 
@@ -5617,6 +5618,11 @@ static int macb_remove(struct platform_device *pdev)
 					  bp->rx_clk, bp->tsu_clk);
 			pm_runtime_set_suspended(&pdev->dev);
 		}
+		pdata = dev_get_platdata(&pdev->dev);
+		if (pdata) {
+			clk_unregister_fixed_rate(pdata->hclk);
+			clk_unregister_fixed_rate(pdata->pclk);
+		}
 		phylink_destroy(bp->phylink);
 		free_netdev(dev);
 	}
-- 
2.47.0

